src/poetry/inspection/lazy_wheel.py

```diff
@@ -0,0 +1,772 @@
+
+    def __enter__(self) -> LazyHTTPFile:
+        """Fetch the remote file length and reset the log of downloaded intervals.
+
+        This method must be called before ``.read()``.
+
+        :raises HTTPRangeRequestUnsupported: if the remote server fails to indicate
+                                             support for "bytes" ranges.
+        """
+        super().__enter__()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> None:
+        """Logs request count to quickly identify any pathological cases in log data."""
+        logger.debug("%d requests for url %s", self._request_count, self._url)
+        super().__exit__(exc_type, exc_value, traceback)
+
+    def _content_length_from_head(self) -> int:
+        """Performs a HEAD request to extract the Content-Length.
+
+        :raises HTTPRangeRequestUnsupported: if the response fails to indicate support
+                                             for "bytes" ranges."""
+        self._request_count += 1
+        head = self._session.head(self._url, headers=self._uncached_headers())
+        head.raise_for_status()
+        assert head.status_code == codes.ok
+        accepted_range = head.headers.get("Accept-Ranges", None)
+        if accepted_range != "bytes":
+            raise HTTPRangeRequestUnsupported(
+                f"server does not support byte ranges: header was '{accepted_range}'"
+            )
+        return int(head.headers["Content-Length"])
+
+    def _stream_response(self, start: int, end: int) -> Response:
+        """Return streaming HTTP response to a range request from start to end."""
+        headers = self._uncached_headers()
+        headers["Range"] = f"bytes={start}-{end}"
+        logger.debug("streamed bytes request: %s", headers["Range"])
+        self._request_count += 1
+        response = self._session.get(self._url, headers=headers, stream=True)
+        response.raise_for_status()
+        assert int(response.headers["Content-Length"]) == (end - start + 1)
+        return response
+
+    @classmethod
+    def _uncached_headers(cls) -> dict[str, str]:
+        """HTTP headers to bypass any HTTP caching.
+
+        The requests we perform in this file are intentionally small, and any caching
+        should be done at a higher level.
+
+        Further, caching partial requests might cause issues:
+        https://github.com/pypa/pip/pull/8716
+        """
+        # "no-cache" is the correct value for "up to date every time", so this will also
+        # ensure we get the most recent value from the server:
+        # https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#provide_up-to-date_content_every_time
+        return {"Accept-Encoding": "identity", "Cache-Control": "no-cache"}
+
+
+class LazyWheelOverHTTP(LazyHTTPFile):
```

Comment:
Do we need all those inheritance levels? From what I have seen, we only use `LazyWheelOverHTTP`. Maybe you could squash some classes together so we can avoid unnecessary code?

Reply:
Good point. I thought the same at first, but then I got used to it because the scope of the classes is clearly separated. Rethinking it, you are probably right that we will never need that many classes. I squashed some of the classes so we now have 3 classes instead of 5. I think that's a good compromise between the number of classes and the length of each class.

I added the changes in separate commits so it's easier to proceed with the review.